src/compiler/python_generator.cc

```diff
@@ -646,42 +644,149 @@ bool PrintPreamble(const FileDescriptor* file,
-      auto service = file->service(i);
-      auto package_qualified_service_name = package + service->name();
-      if (!(PrintStub(package_qualified_service_name, service, &out) &&
-            PrintServicer(service, &out) &&
-            PrintAddServicerToServer(package_qualified_service_name, service,
-                                     &out) &&
-            PrintBetaServicer(service, &out) && PrintBetaStub(service, &out) &&
-            PrintBetaServerFactory(package_qualified_service_name, service,
-                                   &out) &&
-            PrintBetaStubFactory(package_qualified_service_name, service,
-                                 &out))) {
+    Printer out_printer(&output_stream, '$');
+    out = &out_printer;
+
+    if (split_pb2_grpc) {
+      if (!PrintPreambledGAServices()) {
         return make_pair(false, "");
       }
+    } else {
+      out->Print("try:\n");
+      {
+        IndentScope raii_dict_try_indent(out);
+        if (!PrintPreambledGAServices()) {
+          return make_pair(false, "");
+        }
+        if (!PrintBetaServices()) {
+          return make_pair(false, "");
+        }
+      }
+      out->Print("except ImportError:\n");
+      {
+        IndentScope raii_dict_except_indent(out);
+        out->Print("pass");
+      }
     }
   }
   return make_pair(true, std::move(output));
 }
 
+}  // namespace
+
+GeneratorConfiguration::GeneratorConfiguration()
+    : grpc_package_root("grpc"), beta_package_root("grpc.beta") {}
+
+PythonGrpcGenerator::PythonGrpcGenerator(const GeneratorConfiguration& config)
+    : config_(config) {}
+
+PythonGrpcGenerator::~PythonGrpcGenerator() {}
+
+bool PythonGrpcGenerator::Generate(const FileDescriptor* file,
+                                   const grpc::string& parameter,
+                                   GeneratorContext* context,
+                                   grpc::string* error) const {
+  // Get output file name.
+  grpc::string pb2_file_name;
+  grpc::string grpc_file_name;
+  static const int proto_suffix_length = strlen(".proto");
+  if (file->name().size() > static_cast<size_t>(proto_suffix_length) &&
+      file->name().find_last_of(".proto") == file->name().size() - 1) {
+    grpc::string base = file->name().substr(
+        0, file->name().size() - proto_suffix_length);
+    pb2_file_name = base + "_pb2.py";
+    grpc_file_name = base + "_pb2_grpc.py";
+  } else {
+    *error = "Invalid proto file name. Proto file must end with .proto";
+    return false;
+  }
+
+  PrivateGenerator generator(config_, file);
+
+  std::unique_ptr<ZeroCopyOutputStream> pb2_output(
+      context->OpenForAppend(pb2_file_name));
+  std::unique_ptr<ZeroCopyOutputStream> grpc_output(
+      context->Open(grpc_file_name));
+  CodedOutputStream pb2_coded_out(pb2_output.get());
+  CodedOutputStream grpc_coded_out(grpc_output.get());
+  bool success = false;
+  grpc::string pb2_code;
+  grpc::string grpc_code;
+  generator.split_pb2_grpc = false;
```

Comment:
Oh, right: https://github.com/grpc/grpc/pull/8246/files/066fffe0a05401b13b62ed9587ec93d85dca4a66#diff-d5adca4a2d1e65549a73db695e0db419R187

Because there's a lot of shared code between the two up to one tiny difference towards the end and by my not wanting to use lambdas (to make it easier to get to C++98 compatibility) passing around behavior becomes difficult. So instead I used a flag. Totally open to a different suggestion (that preferably doesn't involve adding many new declarations).